tiledb/sm/query/dense_reader.cc

```diff
@@ -0,0 +1,1267 @@
+    case Datatype::DATETIME_HR:
+    case Datatype::DATETIME_MIN:
+    case Datatype::DATETIME_SEC:
+    case Datatype::DATETIME_MS:
+    case Datatype::DATETIME_US:
+    case Datatype::DATETIME_NS:
+    case Datatype::DATETIME_PS:
+    case Datatype::DATETIME_FS:
+    case Datatype::DATETIME_AS:
+    case Datatype::TIME_HR:
+    case Datatype::TIME_MIN:
+    case Datatype::TIME_SEC:
+    case Datatype::TIME_MS:
+    case Datatype::TIME_US:
+    case Datatype::TIME_NS:
+    case Datatype::TIME_PS:
+    case Datatype::TIME_FS:
+    case Datatype::TIME_AS:
+      return dense_read<int64_t, OffType>();
+    default:
+      return LOG_STATUS(Status::ReaderError(
+          "Cannot read dense array; Unsupported domain type"));
+  }
+
+  return Status::Ok();
+}
+
+template <class DimType, class OffType>
+Status DenseReader::dense_read() {
+  // Sanity checks.
+  assert(std::is_integral<DimType>::value);
+
+  // For easy reference.
+  auto& subarray = read_state_.partitioner_.current();
+
+  RETURN_NOT_OK(subarray.compute_tile_coords<DimType>());
+
+  // Compute result space tiles. The result space tiles hold all the
+  // relevant result tiles of the dense fragments.
+  std::map<const DimType*, ResultSpaceTile<DimType>> result_space_tiles;
+  compute_result_space_tiles<DimType>(subarray, &result_space_tiles);
+
+  std::vector<ResultTile*> result_tiles;
+  for (const auto& result_space_tile : result_space_tiles) {
+    for (const auto& result_tile : result_space_tile.second.result_tiles()) {
+      result_tiles.push_back(const_cast<ResultTile*>(&result_tile.second));
+    }
+  }
+
+  // Compute subarrays for each tiles and tile offsets for global order.
+  std::vector<Subarray> tile_subarrays;
+  std::vector<uint64_t> tile_offsets;
+  const auto& tile_coords = subarray.tile_coords();
+  tile_subarrays.reserve(tile_coords.size());
+  tile_offsets.reserve(tile_coords.size());
+
+  const auto& layout =
+      layout_ == Layout::GLOBAL_ORDER ? array_schema_->cell_order() : layout_;
+  uint64_t tile_offset = 0;
+  for (const auto& tc : tile_coords) {
+    tile_subarrays.emplace_back(
+        subarray.crop_to_tile((const DimType*)&tc[0], layout));
+
+    tile_offsets.emplace_back(tile_offset);
+    tile_offset += tile_subarrays.back().cell_num();
+  }
+
+  // Seperate attributes in condition and normal attributes.
+  std::vector<std::string> names(1);
+  std::vector<std::string> condition_names;
+  condition_names.reserve(condition_.field_names().size());
+  std::vector<std::string> attributes_names;
+  attributes_names.reserve(buffers_.size() - condition_.field_names().size());
+
+  for (const auto& it : buffers_) {
+    const auto& name = it.first;
+    if (condition_.field_names().find(name) != condition_.field_names().end()) {
+      names[0] = name;
+      condition_names.emplace_back(name);
+
+      // Pre-load all attribute offsets into memory for attributes
+      // in query condition to be read.
+      load_tile_offsets(&subarray, &names);
```

Comment:
This might be a good idea to perform in parallel, especially for use cases with many attributes and cloud storage. Can be done in a subsequent PR.